samples/net/http_app_server/src/main.c

```diff
@@ -0,0 +1,397 @@
+
+/* Sets the network parameters */
+
+#if defined(CONFIG_NET_CONTEXT_NET_PKT_POOL)
+NET_PKT_TX_SLAB_DEFINE(http_srv_tx, 15);
+NET_PKT_DATA_POOL_DEFINE(http_srv_data, 30);
+
+static struct k_mem_slab *tx_slab(void)
+{
+	return &http_srv_tx;
+}
+
+static struct net_buf_pool *data_pool(void)
+{
+	return &http_srv_data;
+}
+#else
+#if defined(CONFIG_NET_L2_BT)
+#error "TCP connections over Bluetooth need CONFIG_NET_CONTEXT_NET_PKT_POOL "\
+	"defined."
+#endif /* CONFIG_NET_L2_BT */
+
+#define tx_slab NULL
+#define data_pool NULL
+#endif /* CONFIG_NET_CONTEXT_NET_PKT_POOL */
+
+#define RESULT_BUF_SIZE 1500
+static u8_t http_result[RESULT_BUF_SIZE];
+
+#if defined(CONFIG_HTTPS)
+#if !defined(CONFIG_HTTPS_STACK_SIZE)
+#define CONFIG_HTTPS_STACK_SIZE		8192
+#endif /* CONFIG_HTTPS_STACK_SIZE */
+
+#define APP_BANNER "Run HTTPS server"
+#define INSTANCE_INFO "Zephyr HTTPS example server #1"
+
+/* Note that each HTTPS context needs its own stack as there will be
+ * a separate thread for each HTTPS context.
+ */
+NET_STACK_DEFINE(HTTPS, https_stack, CONFIG_NET_APP_TLS_STACK_SIZE,
+		 CONFIG_NET_APP_TLS_STACK_SIZE);
+
+#define RX_FIFO_DEPTH 4
+K_MEM_POOL_DEFINE(ssl_rx_pool, 4, 64, RX_FIFO_DEPTH, 4);
+
+#else /* CONFIG_HTTPS */
+#define APP_BANNER "Run HTTP server"
+#endif /* CONFIG_HTTPS */
+
+/*
+ * Note that the http_server_ctx and http_server_urls are quite large so be
+ * careful if those are allocated from stack.
+ */
+static struct http_ctx http_ctx;
+static struct http_server_urls http_urls;
+
+void panic(const char *msg)
+{
+	if (msg) {
+		NET_ERR("%s", msg);
+	}
+
+	for (;;) {
+		k_sleep(K_FOREVER);
+	}
+}
+
+#define HTTP_STATUS_200_OK	"HTTP/1.1 200 OK\r\n" \
+				"Content-Type: text/html\r\n" \
+				"Transfer-Encoding: chunked\r\n" \
+				"\r\n"
+
+#define HTTP_STATUS_200_OK_GZ	"HTTP/1.1 200 OK\r\n" \
+				"Content-Type: text/html\r\n" \
+				"Transfer-Encoding: chunked\r\n" \
+				"Content-Encoding: gzip\r\n"
+
+#define HTTP_401_STATUS_US	"HTTP/1.1 401 Unauthorized status\r\n" \
+				"WWW-Authenticate: Basic realm=" \
+				"\""HTTP_AUTH_REALM"\"\r\n\r\n"
+
+#define HTML_HEADER		"<html><head>" \
+				"<title>Zephyr HTTP Server</title>" \
+				"</head><body><h1>" \
+				"<center>Zephyr HTTP server</center></h1>\r\n"
+
+#define HTML_FOOTER		"</body></html>\r\n"
+
+static int http_response(struct http_ctx *ctx, const char *header,
+			 const char *payload, size_t payload_len)
+{
+	char content_length[sizeof("Content-Length: ") + 5];
+	struct net_pkt *pkt = NULL;
+	int ret;
+
+	ret = http_add_header(ctx, &pkt, header);
+	if (ret < 0) {
+		NET_ERR("Cannot add HTTP header (%d)", ret);
+		return ret;
+	}
+
+	snprintk(content_length, sizeof(content_length),
+		 "Content-Length: %zd", payload_len);
+	ret = http_add_header(ctx, &pkt, content_length);
+	if (ret < 0) {
+		NET_ERR("Cannot add Content-Length HTTP header (%d)", ret);
+		return ret;
+	}
+
+	http_add_data(ctx, &pkt, HTTP_CRLF, strlen(HTTP_CRLF));
+	http_add_data(ctx, &pkt, payload, payload_len);
```

Comment:
(Looking at what magic this app does that it requires https://github.com/zephyrproject-rtos/zephyr/pull/1330 and whether that can be addressed by https://github.com/zephyrproject-rtos/zephyr/pull/119 pretty easily.)

So, I won't be asking "what happens if payload is `static const payload[1000000]`" like I do in the mailing list RFC, let me just ask "what happens if payload_len is larger than free fragment space available, e.g. payload_len is 3K, but there're only 1.5K of free data buffers." You answer would be "But I allocated 8K of buffers above!" And that's cheating - the only way such paradigm would work is by overallocating buffers to back it.

I don't know where this "browser DOM" approach of "keep everything in memory" comes from. And the problem here is clearly in the API design. It's not natural for an embedded HTTP API to require a user to allocate a "network packet" and then keep adding data to it without limit. The natural API is `http_send_data(context *ctx, const char *buf, int len)`, and let that call decide what to do with that data: accumulate data into infinite buffer (typical "design pattern" for desktop/server apps nowadays), send even the smallest chunk of data right away (the opposite extreme), or buffer it per the natural size of buffer (which is single packet) to minimize copies, when it fills, send it, and repeat with the reminder of data - the approach I advocate in my RFC.

So, it looks like we have a case when "interesting" design of the HTTP API starting to affect design of the IP stack (like requiring #1330).

Reply:
Please note that this is just a sample application that does things in certain way. Nothing prevents developer to send data in certain size chunks here. I just did not implement it here as #1330 was doing it for me automatically.

Reply:
I sent updated HTTP API patches in #4243 
That PR differs slightly from the corresponding patches in this PR, for example the old HTTP API is deprecated, and client API support is added.

Reply:
> Please note that this is just a sample application that does things in certain way. Nothing prevents developer to send data in certain size chunks here. I just did not implement it here as #1330 was doing it for me automatically.

But #1330 is no longer an application patch, it's IP stack patch. If such patches go in, and such sample applications are in the mainline, it sends a strong message how things should be done in Zephyr, nobody simply will look for doing in another way. That's why it's important to have good generic APIs right away and resist desire to push workaround-style solutions down the stack.

Reply:
> I sent updated HTTP API patches in #4243

Thanks, will have a look.